### DISCUSSION TITLE
Optional onchain governance module

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#652] [Ideas]
It would be helpful to give users the ability to optionally attach a governance module to a stream at the time of creation.

The use case is to give users the ability to use the streamed assets for voting.

Ideally, this governance module would connect seamlessly with [Tally](https://www.tally.xyz/).

**Comment by PaulRBerg** [comment]
Had a chat with Matt Solomon from ScopeLift, who are working on a framework for [flexible voting](https://www.scopelift.co/blog/introducing-flexible-voting) in DeFi.

> we're in the process of writing real docs so in the meantime we have a patchwork of blog posts and code, but let me know if you have any questions
> 
> - Intro blog post: https://www.scopelift.co/blog/introducing-flexible-voting
> - Repo: https://github.com/ScopeLift/flexible-voting
> - Governor Implementation: https://github.com/ScopeLift/flexible-voting/blob/master/src/GovernorCountingFractional.sol
> - Voting Client abstract contract, which you can inherit from when building an integration (e.g. sablier would inherit from this), at a minimum a good resource: https://github.com/ScopeLift/flexible-voting/blob/master/src/FlexVotingClient.sol
> - More blog posts under this tag: https://www.scopelift.co/blog/tag/flexible-voting
> 
> and of course it might change based on feedback from users and devs between now and the time you start work on a sablier v3, so any feedback you have (e.g. changes to help you integrate, etc) is appreciated!

**Comment by PaulRBerg** [comment]
Implementing this will be difficult. Broadly speaking, there are two approaches:

1. Flexible voting by inheriting from [`FlexVotingClient`](https://github.com/ScopeLift/flexible-voting/blob/e5de2efd1368387b840931f19f3c184c85842761/src/FlexVotingClient.sol)
2. Deploy a bespoke "vault voting" contract, withdraw the stream deposit there, and reclaim the assets from the vaults when the user wishes to exit the stream. This is what Hedgey does: see [this](https://github.com/hedgey-finance/Locked_VestingTokenPlans/blob/2fdebf08ae8a453e1c7227b44bdfd9133ab4970b/contracts/sharedContracts/VotingVault.sol) and [this](https://github.com/hedgey-finance/Locked_VestingTokenPlans/blob/2fdebf08ae8a453e1c7227b44bdfd9133ab4970b/contracts/LockupPlans/VotingTokenLockupPlans.sol#L458).

The problem with the 1st approach is that it only works with those governance systems that have integrated flexible voting, which AFAIK mean only. Aave, Uniswap, and Gitcoin at the moment.

Whereas the problem with the 2nd approach is .. well, there are multiple problems:

- Significant development cost
- Complicates accounting by moving the funds in a different bespoke contract *per user*
- Gas overhead
    - [410k gas](https://polygonscan.com/tx/0xc92f6984eb209cdf5b01f38c9ad6766da236435a238929b31bed2de9d7a37e28) to create a governance stream on Hedgey
    - [677k gas](https://polygonscan.com/tx/0x0b05b9b763502edb65f0189846f9d315ae929712dc98710483fefd9b3f125c1f) to delegate and deploy a vault contract
    
---

Can you think of any other approaches?

**Comment by PaulRBerg** [comment]
It looks like it might be possible to use Risc0's Bonsai for voting.

[Using SNARKs to enable fully verifiable governance while lowering associated costs](https://twitter.com/RiscZero/status/1717541685767667956).

**Comment by PaulRBerg** [comment]
We should look into potentially supporting OZ's onchain voting mechanism (used by $COMP and $FLUID) since it's the closest to a universal standard in Ethereum.

<img width="400" alt="SCR-20241220-kmzk" src="https://github.com/user-attachments/assets/23541644-e14c-46b7-93ad-3e5d7af14005" />

**Reply by smol-ninja** [reply]
This is the [OZ implementation:](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/utils/Votes.sol)

```solidity
/**
 * @notice Delegate votes from `msg.sender` to `delegatee`
 * @param delegatee The address to delegate votes to
 */
function delegate(address delegatee) public virtual {
    address account = _msgSender();
    _delegate(account, delegatee);
}

/**
 * @dev Delegate all of `account`'s voting units to `delegatee`.
 */
function _delegate(address account, address delegatee) internal virtual {
    address oldDelegate = delegates(account);
    _delegatee[account] = delegatee;

    emit DelegateChanged(account, oldDelegate, delegatee);
    _moveDelegateVotes(oldDelegate, delegatee, _getVotingUnits(account));
}
```
It delegates the entire balance of `msg.sender` to one address. Ideally, for a protocol holding users' tokens, the solution would be to allow partial delegation by choosing an amount in the `delegate` function.

There have been a [few discussions](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3624#issuecomment-1218240345) on it where people have requested similar features. So far, the idea has been rejected as an infeasible solution. Let's say a protocol delegates balance to multiple addresses, then transfers some of the balance. How would the voting contract decide from which delegatee the amount shall be reduced?﻿

-----------

I found a contract in [ENS repo](https://github.com/ensdomains/governance-contracts/blob/master/contracts/ERC20MultiDelegate.sol#L93-L154) with a function to support partial delegations but it seems to be transferring tokens to the contract as a fix to the above problem.

My take is that with the current design of the delegation, we cannot support it until we want to delegate the full balance.

**Reply by PaulRBerg** [reply]
You're right. The OZ model doesn't work when tokens are pooled (like with Sablier).

I guess this is why Scopelift introduced [flexible voting](https://www.scopelift.co/blog/introducing-flexible-voting).

**Comment by PaulRBerg** [comment]
Received a request for this:

<img width="450" alt="SCR-20250305-nitv" src="https://github.com/user-attachments/assets/59a1015a-15fe-4e04-92c6-a6ff97113cf7" />

**Comment by PaulRBerg** [comment]
<img width="300" alt="image" src="https://github.com/user-attachments/assets/fb69b1ad-d64b-4e0f-88e8-95d5a203e843" />

**Comment by PaulRBerg** [comment]
cc @sneg55 

<img width="600" src="https://github.com/user-attachments/assets/b059ff2b-9006-46f6-9a99-11ed74c14554"/>

<img width="600" height="284" alt="image" src="https://github.com/user-attachments/assets/7dd063c2-0e2d-4a96-872f-b00c33990070" />
